streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java

```diff
@@ -388,76 +392,107 @@ public void process(final ProcessorSupplier<K, V> processorSupplier, String... s
+            this.ensureJoinableWith((AbstractStream<K>) other);
+        } else {
+            joinThis.ensureJoinableWith(joinOther);
+        }
 
-        KStreamPassThrough<K, R> joinMerge = new KStreamPassThrough<>();
+        return join.join(joinThis,
+                         joinOther,
+                         joiner,
+                         windows,
+                         keySerde,
+                         thisValueSerde,
+                         otherValueSerde);
+    }
 
-        String thisWindowStreamName = topology.newName(WINDOWED_NAME);
-        String otherWindowStreamName = topology.newName(WINDOWED_NAME);
-        String joinThisName = outer ? topology.newName(OUTERTHIS_NAME) : topology.newName(JOINTHIS_NAME);
-        String joinOtherName = outer ? topology.newName(OUTEROTHER_NAME) : topology.newName(JOINOTHER_NAME);
-        String joinMergeName = topology.newName(MERGE_NAME);
 
-        topology.addProcessor(thisWindowStreamName, thisWindowedStream, this.name);
-        topology.addProcessor(otherWindowStreamName, otherWindowedStream, ((KStreamImpl) other).name);
-        topology.addProcessor(joinThisName, joinThis, thisWindowStreamName);
-        topology.addProcessor(joinOtherName, joinOther, otherWindowStreamName);
-        topology.addProcessor(joinMergeName, joinMerge, joinThisName, joinOtherName);
-        topology.addStateStore(thisWindow, thisWindowStreamName, otherWindowStreamName);
-        topology.addStateStore(otherWindow, thisWindowStreamName, otherWindowStreamName);
+    /**
+     * Repartition a stream. This is required on join operations occurring after
+     * an operation that changes the key, i.e, selectKey, map(..), flatMap(..).
+     * @param keySerde      Serdes for serializing the keys
+     * @param valSerde      Serdes for serilaizing the values
+     * @param copartition   true if we need to co-partition these streams
+     * @return
+     */
+    private KStreamImpl<K, V> repartitionForJoin(Serde<K> keySerde,
+                                                 Serde<V> valSerde,
+                                                 boolean copartition) {
+        String repartitionTopic = this.name + REPARTITION_TOPIC_SUFFIX;
+        String sinkName = topology.newName(SINK_NAME);
+        String sourceName = topology.newName(SOURCE_NAME);
+
+        Serializer<K> keySerializer = keySerde != null ? keySerde.serializer() : null;
+        Serializer<V> valSerializer = valSerde != null ? valSerde.serializer() : null;
+        Deserializer<K> keyDeserializer = keySerde != null ? keySerde.deserializer() : null;
+        Deserializer<V> valDeserializer = valSerde != null ? valSerde.deserializer() : null;
+        topology.addInternalTopic(repartitionTopic);
+        topology.addSink(sinkName, repartitionTopic, keySerializer,
+                                valSerializer, name);
+        topology.addSource(sourceName, keyDeserializer, valDeserializer,
```

Comment:
As I mentioned before, we can do a filtering of null keys here as well since if the key is null, it is going to random partitions and hence meaningless for joins anyways; for example we can add a private `addFilteringSource` in `TopologyBuilder` that filters the null keys. Hence the later filtering on the join operators are only for cases where repartition did not happen before the join.